os/board/esp32_devkit/src/spi_lcd.c

```diff
@@ -0,0 +1,254 @@
+{
+	char buf[4];
+	char devpath[16];
+	int fd;
+	
+	snprintf(devpath, 16, "/dev/gpio%d", port);
+	fd = open(devpath, O_RDWR);
+	if (fd < 0) {
+		lldbg("fd open fail\n");
+		return -1;
+	}
+
+	ioctl(fd, GPIOIOC_SET_DIRECTION, GPIO_DIRECTION_OUT);
+	if (write(fd, buf, snprintf(buf, sizeof(buf), "%d", !!value)) < 0) {
+		lldbg("write error\n");
+	}
+	
+	close(fd);
+	return 0;
+}
+
+void lcd_cmd(spi_device_handle_t spi, const uint8_t cmd, lcd_dc_t *dc)
+{
+	SPI_LOCK(spi, true);
+
+	gpio_set_level((int)dc->dc_io, LCD_CMD_LEV);
+	SPI_SNDBLOCK(spi, &cmd, 1);
+
+	SPI_LOCK(spi, false);
+	return;
+}
+
+void lcd_data(spi_device_handle_t spi, const uint8_t *data, int len, lcd_dc_t *dc)
+{
+	SPI_LOCK(spi, true);
+
+	gpio_set_level((int)dc->dc_io, LCD_DATA_LEV);
+	SPI_SNDBLOCK(spi, data, len);
+
+	SPI_LOCK(spi, false);
+	return;
+}
+
+uint32_t lcd_init(lcd_conf_t *lcd_conf, spi_device_handle_t *spi_wr_dev, lcd_dc_t *dc, int dma_chan)
+{
+	uint32_t lcd_id;
+	int cmd;
+	spi_device_handle_t lcd_handle;
+	const lcd_init_cmd_t *lcd_init_cmds = NULL;
+
+	//Initialize non-SPI GPIOs
+	gpio_pad_select_gpio(lcd_conf->pin_num_dc);
+	gpio_set_direction(lcd_conf->pin_num_dc, GPIO_DIRECTION_OUT);
+
+	//Reset the display
+	if (lcd_conf->pin_num_rst < GPIO_NUM_MAX) {
+		gpio_pad_select_gpio(lcd_conf->pin_num_rst);
+		gpio_set_direction(lcd_conf->pin_num_rst, GPIO_DIRECTION_OUT);
+		gpio_set_level(lcd_conf->pin_num_rst, (lcd_conf->rst_active_level) & 0x1);
+		MSLEEP(100);
+		gpio_set_level(lcd_conf->pin_num_rst, (~(lcd_conf->rst_active_level)) & 0x1);
+		MSLEEP(100);
+	}
+
+	*spi_wr_dev = lcd_handle = up_spiinitialize(lcd_conf->spi_host); //HSPI_PORT
+
+	SPI_SELECT(lcd_handle, CONFIG_ESP32_SPI_CS, true);
+	SPI_SETBITS(lcd_handle, 8);
+	SPI_SETMODE(lcd_handle, SPIDEV_MODE0);
+	SPI_SETFREQUENCY(lcd_handle, lcd_conf->clk_freq);
+	lcd_id = lcd_get_id(lcd_handle, dc);
+	
+	// Use high speed to write LCD
+	SPI_SETFREQUENCY(lcd_handle, lcd_conf->clk_freq);
+
+	if(lcd_conf->lcd_model == LCD_MOD_ST7789) {
+		lcd_init_cmds = st7789_init_cmds;
+	} else if(lcd_conf->lcd_model == LCD_MOD_ILI9341) {
+		lcd_init_cmds = ili_init_cmds;
+	} else if(lcd_conf->lcd_model == LCD_MOD_AUTO_DET) {
+		if (((lcd_id >> 8) & 0xff) == 0x42) {
+			lcd_init_cmds = st7789_init_cmds;
+		} else {
+			lcd_init_cmds = ili_init_cmds;
+		}
+	}
+	assert(lcd_init_cmds != NULL);
+	//Send all the commands
+	cmd = 0;
+	while (lcd_init_cmds[cmd].databytes != 0xff) {
+		lcd_cmd(*spi_wr_dev, lcd_init_cmds[cmd].cmd, dc);
+		lcd_data(*spi_wr_dev, lcd_init_cmds[cmd].data, lcd_init_cmds[cmd].databytes & 0x1F, dc);
+		if (lcd_init_cmds[cmd].databytes & 0x80) {
+			MSLEEP(100);
+		}
+		cmd++;
+	}
+
+	//Enable backlight
+	if (lcd_conf->pin_num_bckl < GPIO_NUM_MAX) {
+		gpio_pad_select_gpio(lcd_conf->pin_num_bckl);
+		gpio_set_direction(lcd_conf->pin_num_bckl, GPIO_DIRECTION_OUT);
+		gpio_set_level(lcd_conf->pin_num_bckl, (lcd_conf->bckl_active_level) & 0x1);
+	}
+
+	lldbg("[lcd_init] finished.\n");

```

Comment:
uilldbg

Reply:
In my opinion, spi_lcd.c is on the same layer with the spiram.c
So spi_lcd should use the lldbg same as spiram.c.